src/server/services/input-types.js

```diff
@@ -115,6 +121,7 @@ function defineType(info) {
 
     types[info.name] = derivedParser;
     typesMeta[info.name] = {
+        service: null,
```

Comment:
I see we handle this in the service loader separately, but it might be nice if `typeTape` was actually a context object like `{service, types}` that included that info already.

Reply:
Yeah, I had thought about making `typeTape` a singleton reference to the current service being loaded (and named something like `currentService`) but...
- If you get a reference to the service in question, you must have already imported it (and defined the types!) so it is too late to set `currentService`
- I like that `withTypeTape` uses an existing pattern (common in python) that leverages RAII. It also allows us to keep `input-types.js` pretty simple rather than starting to more tightly couple it with services themselves by adding a singleton reference to the current service. Minimizing back references in the overall architecture helps keep the project more maintainable than having smaller components know about their parents, etc.

That said, the implementation is a bit quick and dirty. I suspect this would break, too:
```javascript
const [types] = InputTypes.withTypeTape(() => {
  InputTypes.defineType(TYPE_A);
  const [innerTypes] = InputTypes.withTypeTape(() => InputTypes.defineType(TYPE_B));
  InputTypes.defineType(TYPE_C);
});
```
I don't think it is a huge deal since we probably don't need to worry about them being nested just yet but figured I would mention it anyway.

Reply:
Not having a reference to the service makes sense. I was thinking of just having an extra arg for `withTypeTape` which is the service name. So we'd call it (only in the service loader) as `withTypeTape('service-name', () => ...)`. Cause if we don't have the service context (at least the name) at the point of definition, we can't separate the symbol names, so `Color` in PhoneIoT and `Color` in Chart, for instance, would still be a conflict. But if we have at least the service name, we can actually have separate hashmaps for global types (in `input-types.js`) and any particular service's types.

Also, yeah, we probably don't need to worry about nested calls for now, but I just added a tiny change that will make it at least not break by going back to null.

Reply:
Good point about knowing the service when defining the new type to update the collision detection logic. It would be nice to use the official `service.serviceName` rather than the directory name so there isn't another step where the service needs to be mapped to the official name. Anyway, I updated `withTypeTape` so that it doesn't actually register the types defined during the execution of the provided function. This allows the types to be collected and explicitly registered when the name of the service is known.

It currently doesn't handle name collisions of types from different services but I [added a test for it](https://github.com/NetsBlox/NetsBlox/pull/3258/files#diff-aa1a81a07afadf0e041932bfa94bae9a9ffb80a636722d5f6ac913a801e8b471R456-R475) and it can be addressed in a different PR. I am not too worried about including it here since the behavior hasn't changed and, although it is not ideal behavior, it isn't necessarily a hard requirement of this PR. (It will also help keep the PRs smaller which makes reviewing easier, too!).